File: src/main/java/higtools/modules/main/ScaffoldPlus.java
package higtools.modules.main;

import higtools.HIGTools;
import meteordevelopment.meteorclient.events.world.TickEvent;
import meteordevelopment.meteorclient.mixininterface.IVec3d;
import meteordevelopment.meteorclient.settings.*;
import meteordevelopment.meteorclient.systems.modules.Module;
import meteordevelopment.orbit.EventHandler;
import net.minecraft.item.BlockItem;
import net.minecraft.network.packet.c2s.play.PlayerInteractBlockC2SPacket;
import net.minecraft.network.packet.c2s.play.UpdateSelectedSlotC2SPacket;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;

public class ScaffoldPlus extends Module {
    private int slot = -1;
    private boolean worked = false;

    private final SettingGroup sgGeneral = settings.getDefaultGroup();

    private final Setting<Integer> ext = sgGeneral.add(new IntSetting.Builder()
        .name("extend")
        .description("How much to place in front of you.")
        .defaultValue(1)
        .range(0, 5)
        .build()
    );

    private final Setting<Boolean> tower = sgGeneral.add(new BoolSetting.Builder()
        .name("tower")
        .description("Makes towering easier.")
        .defaultValue(false)
        .build()
    );

    private final Setting<Double> towerMult = sgGeneral.add(new DoubleSetting.Builder()
        .name("multi")
        .description("Makes tower potentially bypass stricter anti-cheats.")
        .defaultValue(0.7454)
        .range(0.0, 2.0)
        .visible(tower::get)
        .build()
    );

    private final Setting<Integer> keepY = sgGeneral.add(new IntSetting.Builder()
        .name("keepY")
        .description("Keeps the Y value of the block.")
        .defaultValue(-1)
        .range(-1, 255)
        .sliderRange(-1, 255)
        .build()
    );

    public ScaffoldPlus() {
        super(HIGTools.MAIN, "scaffold+", "Scaffolds blocks under you.");
    }

    @EventHandler
    public void tick(TickEvent.Pre event) {
        float f = MathHelper.sin(mc.player.getYaw() * 0.017453292f);
        float g = MathHelper.cos(mc.player.getYaw() * 0.017453292f);
        int prevSlot = mc.player.getInventory().selectedSlot;

        for (int i = 0; i <= (mc.player.getVelocity().x == 0.0 && mc.player.getVelocity().z == 0.0 ? 0 : ext.get()); i++) {
            // loop body
            Vec3d pos = mc.player.getPos().add(-f * i, -1.0, g * i);
            if (keepY.get() != -1) ((IVec3d) pos).setY(keepY.get() - 1.0);
            BlockPos bpos = BlockPos.ofFloored(pos);
            if (!mc.world.getBlockState(bpos).getMaterial().isReplaceable()) {
                worked = false;
                continue;
            }
            worked = true;

            boolean offHand = mc.player.getOffHandStack().getItem() instanceof BlockItem;
            boolean mainHand = mc.player.getMainHandStack().getItem() instanceof BlockItem;
            if (!offHand && !mainHand) {
                for (int j = 0; j <= 8; j++) {
                    if (mc.player.getInventory().getStack(j).getItem() instanceof BlockItem) {
                        slot = j;
                        break;
                    }
                }
                if (slot == -1) return;
                mc.player.getInventory().selectedSlot = slot;
                mc.player.networkHandler.sendPacket(new UpdateSelectedSlotC2SPacket(slot));
            }

            if (tower.get() && mc.options.jumpKey.isPressed() && mc.player.getVelocity().x == 0.0 && mc.player.getVelocity().z == 0.0) {
                if (mc.world.getBlockState(mc.player.getBlockPos().down()).getMaterial().isReplaceable() &&
                    !mc.world.getBlockState(mc.player.getBlockPos().down(2)).getMaterial().isReplaceable() &&
                    mc.player.getVelocity().y > 0) {
                    mc.player.setVelocity(mc.player.getVelocity().x, -0.6, mc.player.getVelocity().z);
                    mc.player.jump();
                    mc.player.setVelocity(mc.player.getVelocity().x, mc.player.getVelocity().y * towerMult.get(), mc.player.getVelocity().z);
                }
            }

            mc.player.networkHandler.sendPacket(
                new PlayerInteractBlockC2SPacket(
                    offHand ? Hand.OFF_HAND : Hand.MAIN_HAND,
                    new BlockHitResult(pos, Direction.DOWN, bpos, false),
                    0
                )
            );
            slot = -1;
        }

        if (mc.player.getInventory().selectedSlot != prevSlot) {
            mc.player.getInventory().selectedSlot = prevSlot;
            mc.player.networkHandler.sendPacket(new UpdateSelectedSlotC2SPacket(prevSlot));
        }
    }

    public boolean isWorked() {
        return worked;
    }
}

File: src/main/kotlin/higtools/modules/kmain/ScaffoldPlus.kt
package higtools.modules.kmain

import higtools.*
import higtools.HIGTools
import meteordevelopment.meteorclient.events.world.TickEvent
import meteordevelopment.meteorclient.mixininterface.IVec3d
import meteordevelopment.orbit.EventHandler
import net.minecraft.item.BlockItem
import net.minecraft.network.packet.c2s.play.PlayerInteractBlockC2SPacket
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.math.MathHelper
import net.minecraft.network.packet.c2s.play.UpdateSelectedSlotC2SPacket

object ScaffoldPlus:MeteorModule(HIGTools.MAIN, "Scaffold+", "Scaffolds blocks under you.") {

    private var slot = -1

    private var ext by mainGroup.add(IValue("Extend", 1, "How much to place in front of you", 0..5, 1))
    private var tower by mainGroup.add(BValue("Tower", false, "Makes towering easier"))
    private var towerMult by mainGroup.add(DValue("Multi", 0.7454, "Makes tower potentially bypass stricter anti-cheats.", 0.0..2.0, 0.0001))
    private var keepY by mainGroup.add(IValue("KeepY", -1, "Keeps the Y value of the block", -1..255, 1))

    var worked:Boolean = false

    @EventHandler
    fun tick(event:TickEvent.Pre) {
        val f = MathHelper.sin(mc.player!!.yaw * 0.017453292f)
        val g = MathHelper.cos(mc.player!!.yaw * 0.017453292f)
        var runs = 0
        val prevSlot = mc.player!!.inventory.selectedSlot
        for (i in 0..if (mc.player!!.velocity.x == 0.0 && mc.player!!.velocity.z == 0.0) 0 else ext) {
            val pos = mc.player!!.pos.add(-f * i.toDouble(), -1.0, g * i.toDouble())
            if (keepY != -1) {
                (pos as IVec3d).setY(keepY.toDouble() - 1.0)
            }
            val bpos = BlockPos(pos)
            if (!mc.world!!.getBlockState(bpos).material.isReplaceable) {
                worked = false
                continue
            }
            worked = true
            val offHand = mc.player!!.offHandStack.item is BlockItem
            if (!offHand) {
                if (mc.player!!.mainHandStack.item !is BlockItem) {
                    findBlock@
                    for (j in 0..8) {
                        if (mc.player!!.inventory.getStack(j).item is BlockItem) {
                            slot = j
                            break@findBlock
                        }
                    }
                    if (slot == -1) return
                    mc.player!!.inventory.selectedSlot = slot
                    mc.networkHandler!!.sendPacket(UpdateSelectedSlotC2SPacket(slot))
                }
            }
            if (tower && mc.options.jumpKey.isPressed && mc.player!!.velocity.x == 0.0 && mc.player!!.velocity.z == 0.0) {
                if (mc.world!!.getBlockState(mc.player!!.blockPos.down()).material.isReplaceable && !mc.world!!.getBlockState(
                        mc.player!!.blockPos.down(2)).material.isReplaceable && mc.player!!.velocity.y > 0) {
                    mc.player!!.setVelocity(mc.player!!.velocity.x, -0.6, mc.player!!.velocity.z)
                    mc.player!!.jump()
                    mc.player!!.setVelocity(mc.player!!.velocity.x,
                                            mc.player!!.velocity.y * towerMult,
                                            mc.player!!.velocity.z)
                }
            }
            mc.networkHandler!!.sendPacket(
                PlayerInteractBlockC2SPacket(
                    if (offHand) Hand.OFF_HAND else Hand.MAIN_HAND,
                    BlockHitResult(pos, Direction.DOWN, bpos, false), 0))
            runs++
            slot = -1
        }
        if (mc.player!!.inventory.selectedSlot != prevSlot) {
            mc.player!!.inventory.selectedSlot = prevSlot
            mc.networkHandler!!.sendPacket(UpdateSelectedSlotC2SPacket(prevSlot))
        }
    }

}
